src/sctools/fastq.py

```diff
@@ -0,0 +1,197 @@
+from collections import namedtuple
+from . import reader
+
+
+class Record:
+    """Abstract Fastq record object
+
+    Defines properties for accessing fastq record information:
+    :property name: name field
+    :property sequence: sequence field
+    :property name2: second name field
+    :property quality: quality field
+
+    Utilities for parsing read attributes:
+    :property average_quality: return the mean quality of FastqRecord
+    """
+    __slots__ = ['_data']
+
+    def __init__(self, record):
+        """
+        :param [str|bytes] record: list of four strings or bytes objects defining a
+          single fastq record
+        """
+        self._data = list(record)
+
+    @property
+    def name(self):
+        return self._data[0]
+
+    @name.setter
+    def name(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('name must be str or bytes')
+        else:
+            self._data[0] = value
+
+    @property
+    def sequence(self):
+        return self._data[1]
+
+    @sequence.setter
+    def sequence(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('sequence must be str or bytes')
+        else:
+            self._data[1] = value
+
+    @property
+    def name2(self):
+        return self._data[2]
+
+    @name2.setter
+    def name2(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('name2 must be str or bytes')
+        else:
+            self._data[2] = value
+
+    @property
+    def quality(self):
+        return self._data[3]
+
+    @quality.setter
+    def quality(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('quality must be str or bytes')
+        else:
+            self._data[3] = value
+
+    def __bytes__(self):
+        try:
```

Comment:
Do you expect _data to usually be in bytes, and/or do you prefer to optimize speed for the bytes case, rather than for string? That appears to be the choice you're making here, just want to confirm.

Reply:
Would it be worth moving implementation of __bytes__ and __str__ into the BytesRecord and StrRecord classes, so you don't have to try catch?

Reply:
I expect it to almost always be bytes (and it will always be bytes for our pipelines), as
it's faster to convert from bytes to 2-bit encodings; which you will see later in the package. 

I used to have these inside the respective classes as you're suggesting. The reason I moved away from it was essentially just to reduce the amount of code in the package. My logic was that anyone using strings no longer cares about speed, and therefore it's OK to add small inefficiencies for string records. 

I am very willing to move it back if you feel that would be better!

Reply:
.. I will also make a note to be clear bytes is the preferred implementation.

Reply:
It's such a small amount of code and you have to have it somewhere - seems a little clearer to put it in BytesRecord and StrRecord.

Reply:
Adjusted to make bytes and str both 1st class. StrRecord now subclasses Record, which operates on bytes. This should also indicate to the user (subtly) that bytes is the preferred implementation.